# HG changeset patch
# User Rik <rik@octave.org>
# Date 1543349681 28800
#      Tue Nov 27 12:14:41 2018 -0800
# Node ID b9d72a2dac8f1cedcf0a823363081792c5a870df
# Parent  42a06f8e69662e7fe4f93c042490acde2e86d115
Accept a cell array of strings to uncompress functions (bug #55102).

* bunzip2.m, gunzip.m: Document third calling form which returns FILELIST.
For single file arguments, extract the destination dir from filename.

* unpack.m: Change default for input DIR to [].
For cell array input, recursively call unpack and supply either the specified
DIR, or get the output DIR from the filename.

* untar.m, unzip.m: Document third calling form which returns FILELIST.

diff --git a/scripts/miscellaneous/bunzip2.m b/scripts/miscellaneous/bunzip2.m
--- a/scripts/miscellaneous/bunzip2.m
+++ b/scripts/miscellaneous/bunzip2.m
@@ -12,18 +12,19 @@
 ## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 ## GNU General Public License for more details.
 ##
 ## You should have received a copy of the GNU General Public License
 ## along with Octave; see the file COPYING.  If not, see
 ## <https://www.gnu.org/licenses/>.
 
 ## -*- texinfo -*-
-## @deftypefn  {} {@var{filelist} =} bunzip2 (@var{bzfile})
-## @deftypefnx {} {@var{filelist} =} bunzip2 (@var{bzfile}, @var{dir})
+## @deftypefn  {} {} bunzip2 (@var{bzfile})
+## @deftypefnx {} {} bunzip2 (@var{bzfile}, @var{dir})
+## @deftypefnx {} {@var{filelist} =} bunzip2 (@dots{})
 ## Unpack the bzip2 archive @var{bzfile}.
 ##
 ## If @var{dir} is specified the files are unpacked in this directory rather
 ## than the one where @var{bzfile} is located.
 ##
 ## The optional output @var{filelist} is a list of the uncompressed files.
 ## @seealso{bzip2, unpack, gunzip, unzip, untar}
 ## @end deftypefn
@@ -31,17 +32,17 @@
 ## Author: Bill Denney <denney@seas.upenn.edu>
 
 function filelist = bunzip2 (bzfile, dir = [])
 
   if (nargin < 1 || nargin > 2)
     print_usage ();
   endif
 
-  if (isempty (dir))
+  if (isempty (dir) && ischar (bzfile))
     dir = fileparts (bzfile);
   endif
 
   if (nargout > 0)
     filelist = unpack (bzfile, dir, "bz2");
   else
     unpack (bzfile, dir, "bz2");
   endif
diff --git a/scripts/miscellaneous/gunzip.m b/scripts/miscellaneous/gunzip.m
--- a/scripts/miscellaneous/gunzip.m
+++ b/scripts/miscellaneous/gunzip.m
@@ -12,18 +12,19 @@
 ## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 ## GNU General Public License for more details.
 ##
 ## You should have received a copy of the GNU General Public License
 ## along with Octave; see the file COPYING.  If not, see
 ## <https://www.gnu.org/licenses/>.
 
 ## -*- texinfo -*-
-## @deftypefn  {} {@var{filelist} =} gunzip (@var{gzfile})
-## @deftypefnx {} {@var{filelist} =} gunzip (@var{gzfile}, @var{dir})
+## @deftypefn  {} {} gunzip (@var{gzfile})
+## @deftypefnx {} {} gunzip (@var{gzfile}, @var{dir})
+## @deftypefnx {} {@var{filelist} =} gunzip (@dots{})
 ## Unpack the gzip archive @var{gzfile}.
 ##
 ## If @var{gzfile} is a directory, all gzfiles in the directory will be
 ## recursively unpacked.
 ##
 ## If @var{dir} is specified the files are unpacked in this directory rather
 ## than the one where @var{gzfile} is located.
 ##
@@ -34,17 +35,17 @@
 ## Author: Bill Denney <denney@seas.upenn.edu>
 
 function filelist = gunzip (gzfile, dir = [])
 
   if (nargin < 1 || nargin > 2)
     print_usage ();
   endif
 
-  if (isempty (dir))
+  if (isempty (dir) && ischar (gzfile))
     dir = fileparts (gzfile);
   endif
 
   if (nargout > 0)
     filelist = unpack (gzfile, dir, "gz");
   else
     unpack (gzfile, dir, "gz");
   endif
diff --git a/scripts/miscellaneous/unpack.m b/scripts/miscellaneous/unpack.m
--- a/scripts/miscellaneous/unpack.m
+++ b/scripts/miscellaneous/unpack.m
@@ -66,17 +66,17 @@
 ## @end table
 ##
 ## The optional return value is a list of @var{files} unpacked.
 ## @seealso{bunzip2, gunzip, unzip, untar, bzip2, gzip, zip, tar}
 ## @end deftypefn
 
 ## Author: Bill Denney <denney@seas.upenn.edu>
 
-function filelist = unpack (file, dir = ".", filetype = "")
+function filelist = unpack (file, dir = [], filetype = "")
 
   if (nargin < 1 || nargin > 3)
     print_usage ();
   endif
 
   if (! ischar (file) && ! iscellstr (file))
     error ("unpack: FILE must be a string or cell array of strings");
   endif
@@ -96,17 +96,21 @@ function filelist = unpack (file, dir = 
       endif
     endif
   endif
 
   ## Recursively unpack cellstr arrays one file at a time
   if (numel (file) > 1)
     files = {};
     for i = 1:numel (file)
-      tmpfiles = unpack (file{i}, dir);
+      if (! isempty (dir))
+        tmpfiles = unpack (file{i}, dir);
+      else
+        tmpfiles = unpack (file{i}, fileparts (file{i}));
+      endif
       files = {files{:} tmpfiles{:}};
     endfor
 
     ## Return output if requested.
     if (nargout > 0)
       filelist = files;
     endif
     return;
diff --git a/scripts/miscellaneous/untar.m b/scripts/miscellaneous/untar.m
--- a/scripts/miscellaneous/untar.m
+++ b/scripts/miscellaneous/untar.m
@@ -14,16 +14,17 @@
 ##
 ## You should have received a copy of the GNU General Public License
 ## along with Octave; see the file COPYING.  If not, see
 ## <https://www.gnu.org/licenses/>.
 
 ## -*- texinfo -*-
 ## @deftypefn  {} {} untar (@var{tarfile})
 ## @deftypefnx {} {} untar (@var{tarfile}, @var{dir})
+## @deftypefnx {} {@var{filelist} =} untar (@dots{})
 ## Unpack the TAR archive @var{tarfile}.
 ##
 ## If @var{dir} is specified the files are unpacked in this directory rather
 ## than the one where @var{tarfile} is located.
 ##
 ## The optional output @var{filelist} is a list of the uncompressed files.
 ## @seealso{tar, unpack, bunzip2, gunzip, unzip}
 ## @end deftypefn
diff --git a/scripts/miscellaneous/unzip.m b/scripts/miscellaneous/unzip.m
--- a/scripts/miscellaneous/unzip.m
+++ b/scripts/miscellaneous/unzip.m
@@ -12,18 +12,19 @@
 ## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 ## GNU General Public License for more details.
 ##
 ## You should have received a copy of the GNU General Public License
 ## along with Octave; see the file COPYING.  If not, see
 ## <https://www.gnu.org/licenses/>.
 
 ## -*- texinfo -*-
-## @deftypefn  {} {@var{filelist} =} unzip (@var{zipfile})
-## @deftypefnx {} {@var{filelist} =} unzip (@var{zipfile}, @var{dir})
+## @deftypefn  {} {} unzip (@var{zipfile})
+## @deftypefnx {} {} unzip (@var{zipfile}, @var{dir})
+## @deftypefnx {} {@var{filelist} =} unzip (@dots{})
 ## Unpack the ZIP archive @var{zipfile}.
 ##
 ## If @var{dir} is specified the files are unpacked in this directory rather
 ## than the one where @var{zipfile} is located.
 ##
 ## The optional output @var{filelist} is a list of the uncompressed files.
 ## @seealso{zip, unpack, bunzip2, gunzip, untar}
 ## @end deftypefn
